pkg/kubelet/network/exec/exec.go

```
@@ -0,0 +1,102 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package exec
+
+import (
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path"
+
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/kubelet/dockertools"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/kubelet/network"
+	cmdexec "github.com/GoogleCloudPlatform/kubernetes/pkg/util/exec"
+	"github.com/golang/glog"
+)
+
+type execNetworkPlugin struct {
+	execName string
+	execPath string
+	host     network.Host
+}
+
+const (
+	initCmd     = "--init"
+	setUpCmd    = "--setup"
+	tearDownCmd = "--teardown"
+	execDir     = "/usr/libexec/kubernetes/kubelet-plugins/net/exec/"
+)
+
+func ProbeNetworkPlugins() []network.NetworkPlugin {
+	return ProbeNetworkPluginsWithExecDir(execDir)
+}
+
+func ProbeNetworkPluginsWithExecDir(pluginDir string) []network.NetworkPlugin {
+	execPlugins := []network.NetworkPlugin{}
+
+	files, _ := ioutil.ReadDir(pluginDir)
+	for _, f := range files {
+		// only directories are counted as plugins
+		// and we expect pluginDir/dirname/dirname to be an executable
+		if f.IsDir() {
+			execPath := path.Join(pluginDir, f.Name())
+			execPlugins = append(execPlugins, &execNetworkPlugin{execName: f.Name(), execPath: execPath})
+		}
+	}
+	return execPlugins
+}
+
+func (plugin *execNetworkPlugin) Init(host network.Host) error {
+	plugin.host = host
+	// call the init script
+	executable := path.Join(plugin.execPath, plugin.execName)
+	out, err := cmdexec.New().Command(executable, initCmd).CombinedOutput()
+	glog.V(5).Infof("Init 'exec' network plugin output: %s, %v", string(out), err)
+	return err
+}
+
+func (plugin *execNetworkPlugin) Name() string {
+	return plugin.execName
+}
+
+func (plugin *execNetworkPlugin) Validate() error {
```

I wanted plugins to do be able to check whether they are installed properly or not. Helps separating the code that should fail fast. Not stuck on it.